### PR TITLE
[gui][server] Statistics

### PR DESCRIPTION
**Pull request by csordasmarton**
- Show statistics in separate tabs.
- Component statistics.
- Compare statistics.
- Improve compare queries by using subqueries.

### CODE REVIEW

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 41
           

Complexity increasing per file
==============================
- web/server/vue-cli/src/components/Statistics/StatisticsHelper.js  3
         

Complexity decreasing per file
==============================
+ web/server/codechecker_server/api/report_server.py  -1
         

Clones removed
==============
+ web/server/codechecker_server/api/report_server.py  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=6038424&bid=19786675)

File: web/server/codechecker_server/api/report_server.py [hunk 1]
```diff
@@ -349,6 +347,81 @@ def process_report_filter(session, report_filter):
     return filter_expr
 
 
+def get_outstanding_reports_filter(tbl=Report):
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py:350]: this was renamed to open reports right?

what is the `tbl=Report` expression, Report is a type right?

File: web/server/codechecker_server/api/report_server.py [hunk 2]
```diff
@@ -349,6 +347,81 @@ def process_report_filter(session, report_filter):
...
+                        RunHistory.run_id == Run.id) \
+             .filter(RunHistory.id.in_(tag_ids))
+
+    return q
+
+
+def process_cmp_data_filter(session, run_ids, report_filter, cmp_data):
+    """ Process compare data filter. """
+    base_tag_ids = report_filter.runTag if report_filter else None
+    query_a = get_diff_bug_id_query(session, run_ids, base_tag_ids)
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py:390]: could you add a bit more meaningful names to query_a ans query_b, one of them is the base and the other is the new?

File: web/server/codechecker_server/api/report_server.py [hunk 3]
```diff
@@ -563,20 +610,13 @@ def create_count_expression(report_filter):
         return func.count(literal_column('*'))
 
 
-def filter_report_filter(q, filter_expression, run_ids=None, cmp_data=None,
-                         diff_hashes=None):
-    if run_ids:
-        q = q.filter(Report.run_id.in_(run_ids))
-
+def filter_report_filter(q, filter_expression):
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py:613]: Maybe we should rename this function to something more descriptive and add a docstring about why it is needed.
It just joins the File and ReviewStatus tables and applies the filter expression.

File: web/server/codechecker_server/api/report_server.py [hunk 4]
```diff
@@ -2125,10 +1996,7 @@ def getRunHistoryTagCounts(self, run_ids, report_filter, cmp_data, limit,
                 .outerjoin(report_cnt_q,
                            report_cnt_q.c.run_id == RunHistory.run_id) \
                 .filter(RunHistory.version_tag.isnot(None)) \
-                .filter(and_(report_cnt_q.c.detected_at <= RunHistory.time,
-                             or_(report_cnt_q.c.fixed_at.is_(None),
-                                 report_cnt_q.c.fixed_at >=
-                                 RunHistory.time))) \
+                .filter(get_outstanding_reports_filter(report_cnt_q.c)) \
```
Comment (gyorb) [inline web/server/codechecker_server/api/report_server.py:1999]: open reports filter? Shouldn't the query from the other PR be used when that is merged?

**Review by gyorb — CHANGES_REQUESTED** [review]
The default sorting could be based on the severity of the checkers in the component statistics view too.

Also please check travis, the tests did not run.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 41
           

Complexity increasing per file
==============================
- web/server/vue-cli/src/components/Statistics/StatisticsHelper.js  3
         

Complexity decreasing per file
==============================
+ web/server/codechecker_server/api/report_server.py  -1
         

Clones removed
==============
+ web/server/codechecker_server/api/report_server.py  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=6038424&bid=19786675)

**Review by gyorb — CHANGES_REQUESTED** [review]
please resolve the merge conflict

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 41
           

Complexity increasing per file
==============================
- web/server/vue-cli/src/components/Statistics/StatisticsHelper.js  3
         

Complexity decreasing per file
==============================
+ web/server/codechecker_server/api/report_server.py  -2
         

Clones removed
==============
+ web/server/codechecker_server/api/report_server.py  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=6038424&bid=19786675)

**Review by gyorb — CHANGES_REQUESTED** [review]
I've tested some use cases and found some interesting behavior.

click on the statistics/component statistics tab and there is let's say 2 confirmed bugs for a component.
click on the number 2
the UI jumps to the reports tab but the reports are not filtered based on the checker name the review status and the source component, I see much more results not just the 2 which were expected.
Is this behavior known and will be resolved in #2913?

**Comment by csordasmarton** [comment]
@gyorb Yes, this problem will be solved in #2913.

**Review by gyorb — APPROVED** [review]

**Pull request merged**